### PR TITLE
Support permitted subclasses

### PR DESCRIPTION
**Pull request by pkoenig10**

### CODE REVIEW

File: javapoet/src/main/java/com/palantir/javapoet/TypeSpec.java [hunk 1]
```diff
@@ -669,6 +686,51 @@ public Builder addSuperinterface(TypeMirror superinterface, boolean avoidNestedT
             return this;
         }
 
+        public Builder addPermittedSubclasses(Iterable<? extends TypeName> permittedSubclasses) {
```
Comment (pkoenig10) [inline javapoet/src/main/java/com/palantir/javapoet/TypeSpec.java:694]: Chose this name to match [`Class.getPermittedSubclasses`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Class.html#getPermittedSubclasses).

**Review by carterkozak — COMMENTED** [review]
I think we're missing a test-case for explicitly `non-sealed` classes, which would currently generate as `non_sealed class Foo` instead of `non-sealed class Foo`.

There are a couple forks of javapoet where folks have implemented this feature:
https://github.com/FabricMC/javapoet/pull/3/files
https://github.com/Randgalt/java-composer/pull/3/files

File: javapoet/src/main/java/com/palantir/javapoet/TypeSpec.java (hunk 1 above)
Comment (carterkozak) [reply javapoet/src/main/java/com/palantir/javapoet/TypeSpec.java:694]: Makes sense 👍

**Comment by carterkozak** [comment]
We'll probably want to add support for records as well before too long

**Comment by pkoenig10** [comment]
I didn't bother adding checks to disallow `SEALED` and `NON_SEALED` modifiers for fields or methods like in https://github.com/Randgalt/java-composer/pull/3. There are a bunch of other invalid modifiers for which we don't have checks today. In practice the value of these checks is relatively low - the generated code won't compile so it's pretty obvious when this goes wrong.

**Comment by carterkozak** [comment]
👍 agreed, as long as we're generating the right thing, I'm happy. Generating code that doesn't compile is likely easier to debug than throwing a runtime exception in most cases, and we don't want to add restrictions that may or may not align with future java compilers.

**Review by carterkozak — APPROVED** [review]

**Pull request merged**